### PR TITLE
Consume Span moves

### LINKED ISSUE #24854: Provide IEnumerable<T> support for Memory<T>
## Rationale
Users should be able to pass `Memory<T>` to methods that take in `IEnumerable<T>` similar to array. Currently, this requires calling ToArray() to copy the contents of the `Memory<T>` into an array for it to work, and is unnecessary.

For example (from @eerhardt):
https://github.com/dotnet/corefxlab/blob/master/src/System.Numerics.Tensors/System/Numerics/DenseTensor.cs#L10
```C#
public void SomeMethod(DenseTensor<float> tensor)
{
   Memory<float> memory = tensor.Buffer;
   var value = CreateBatch<T>(..., memory.ToArray(), ...);
}
```

https://github.com/Microsoft/CNTK/blob/master/bindings/csharp/CNTKLibraryManagedDll/ShimApiClasses/ValueShim.cs#L104

```C#
public static Value CreateBatch<T>(NDShape sampleShape, IEnumerable<T> batch, DeviceDescriptor device, bool readOnly = false)
{
    T[] batchAsArray = batch.ToArray();
    return CreateBatch<T>(sampleShape, batchAsArray, 0, batchAsArray.Count(), device, readOnly);
}
```

If `Memory<T>` implemented `IEnumerable<T>`, then the ToArray() call would not be required. However, this would cause applications that reference System.Memory and `Memory<T>` for any scenario to be larger (when compiled AOT). Furthermore, if `Memory<T>` was an `IEnumerable<T>`, it might result in a pit of failure as it could lead to users unintentionally using the enumerator to iterate over the data rather than the more performant indexer on Memory.Span (especially for primitive types like `Memory<byte>`). To discourage unnecessary use of the enumerator but still provide support for the scenarios where you need an IEnumerable, the proposed solution is to add an adapter class and a ToEnumerable extension method on `Memory<T>`.

As an FYI, `Span<T>` cannot implement `IEnumerable<T>` since it is a stack-only, byref type, and casting it to an interface would cause boxing. The compiler will throw and error: `'Span<T>': ref structs cannot implement interfaces`

## Proposed API
```C#
namespace System
{
   public static class MemoryExtensions {
     public static IEnumerable<T> ToEnumerable(this Memory<T> memory);
   }
}
```

## Usage
Taking the above example, it would look as follows:
```C#
public void SomeMethod(DenseTensor<float> tensor)
{
   Memory<float> memory = tensor.Buffer;
   var value = CreateBatch<T>(..., memory.ToEnumerable(), ...);
}
```

## Partial Implementation
```C#
public static IEnumerable<T> ToEnumerable(this Memory<T> memory) => new MemoryEnumerable<T>(memory);

internal class MemoryEnumerable<T> : IEnumerable<T>
{
    ReadOnlyMemory<T> _memory;

    public MemoryEnumerable(ReadOnlyMemory<T> memory) => _memory = memory;

    public IEnumerator<T> GetEnumerator() => new MemoryEnumerator<T>(_memory);

    IEnumerator IEnumerable.GetEnumerator() => new MemoryEnumerator<T>(_memory);
}

internal class MemoryEnumerator<T> : IEnumerator<T>
{
    ReadOnlyMemory<T> _memory;
    int _index;

    public MemoryEnumerator(ReadOnlyMemory<T> memory) => _memory = memory;

    public T Current => _memory.Span[_index];

    object IEnumerator.Current => _memory.Span[_index];

    public void Dispose()
    {
        throw new NotImplementedException();
    }

    public bool MoveNext()
    {
        _index++;
        return _index <_memory.Length;
    }

    public void Reset()
    {
        _index = 0;
    }
}

```

cc @eerhardt, @KrzysztofCwalina, @stephentoub, @jkotas, @terrajobst, @karelz, @ericstj

### LINKED ISSUE #26139: Move Span.DangerousCreate to MemoryMarshal.Create
current API:
```c#
    public struct Span<T> {
        public static Span<T> DangerousCreate(object obj, ref T objectData, int length);
    }
    public struct ReadOnlySpan<T> {
        public static ReadOnlySpan<T> DangerousCreate(object obj, ref T objectData, int length);
    }
```

  proposed API:
```c#
    public static class MemoryMarshal {
        public static Span<T> CreateSpan(object obj, ref T objectData, int length);
        public static ReadOnlySpan<T> CreateReadOnlySpan(object obj, ref T objectData, int length);
    }
```

### LINKED ISSUE #26295: Span: Provide ToString
Forked from https://github.com/dotnet/corefx/issues/14376

Work item here is to implement ToString on Span/ReadOnlySpan and have it print out a string similar to what a collection would print out (which is just the object tostring - type + address)

cc: @KrzysztofCwalina @jkotas @ahsonkhan to continue the discussion on ToString here.

### LINKED ISSUE #26368: Move Span.NonPortableCast to MemoryMarshal and rename to Cast
From https://github.com/dotnet/corefx/issues/26139#issuecomment-358090527:

Move and rename NonPortableCast to MemoryMarshal.Cast

**Current API:**
```C#
namespace System
{
    public static partial class MemoryExtensions
    {
        public static ReadOnlySpan<TTo> NonPortableCast<TFrom, TTo>(this ReadOnlySpan<TFrom> source) where TFrom : struct where TTo : struct { throw null; }
        public static Span<TTo> NonPortableCast<TFrom, TTo>(this Span<TFrom> source) where TFrom : struct where TTo : struct { throw null; }
    }
}
```

**Proposed API:**
```C#
namespace System.Runtime.InteropServices
{
    public static partial class MemoryMarshal
    {
        public static bool TryCast<TFrom, TTo>(ReadOnlySpan<TFrom> source, out ReadOnlySpan<TTo> output) where TFrom : struct where TTo : struct { throw null; }
        public static bool TryCast<TFrom, TTo>(Span<TFrom> source, out Span<TTo> output) where TFrom : struct where TTo : struct { throw null; }
    }
}
```

**Edit:** Removed `this`
**Edit** by @ianhays: Changed `Cast` to `TryCast` per discussion

cc @jkotas, @KrzysztofCwalina

### PR DESCRIPTION
**Pull request by ianhays**
- [Move Span.DangerousCreate to MemoryMarshal.CreateSpan](https://github.com/dotnet/corefx/issues/26139)
- [Move ReadOnlySpan.DangerousCreate to MemoryMarshal.CreateReadOnlySpan](https://github.com/dotnet/corefx/issues/26139)
- [Move Span.NonPortableCast to MemoryMarshal.Cast](https://github.com/dotnet/corefx/issues/26368)
- [Move ReadOnlySpan.NonPortableCast to MemoryMarshal.Cast](https://github.com/dotnet/corefx/issues/26368)
- [Add ToString override to Span and ReadOnlySpan](https://github.com/dotnet/corefx/issues/26295)
- [Add ToEnumerable function to MemoryMarshal that takes a Memory](https://github.com/dotnet/corefx/issues/24854)
- Since the above require adding a fast-span specific API (the new Create functions), I had to add netcoreapp and uap configurations to the test build so we only run the Create tests against the fast-span.

Depends on https://github.com/dotnet/coreclr/pull/15941

resolves #26368 
resolves #24854
resolves #26295 
resolves #26139

### CODE REVIEW

File: src/System.Memory/ref/System.Memory.csproj [hunk 1]
```diff
@@ -7,6 +7,7 @@
     <!-- disable warnings about obsolete APIs -->
     <NoWarn>$(NoWarn);0809</NoWarn>
     <IsPartialFacadeAssembly Condition="'$(TargetGroup)' == 'netcoreapp' Or '$(TargetGroup)' == 'uap'">true</IsPartialFacadeAssembly>
+    <DefineConstants Condition="'$(IsPartialFacadeAssembly)' != 'true'">$(DefineConstants);netstandard;FEATURE_PORTABLE_SPAN</DefineConstants>
```
Comment (ahsonkhan) [inline src/System.Memory/ref/System.Memory.csproj]: Can we remove the netstandard constant?

File: src/System.Memory/tests/MemoryMarshal/CreateReadOnlySpan.cs [hunk 2]
```diff
@@ -10,10 +10,10 @@
 
 namespace System.SpanTests
 {
-    public static partial class ReadOnlySpanTests
+    public static partial class MemoryMarshalTests
     {
         [Fact]
-        public static void DangerousCreate1()
+        public static void CreateReadOnlySpan1()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/MemoryMarshal/CreateReadOnlySpan.cs]: nit: remove the `1` suffix from the test name.

File: src/System.Memory/tests/MemoryMarshal/CreateSpan.cs [hunk 3]
```diff
@@ -10,10 +10,10 @@
 
 namespace System.SpanTests
 {
-    public static partial class SpanTests
+    public static partial class MemoryMarshalTests
     {
         [Fact]
-        public static void DangerousCreate1()
+        public static void CreateSpan1()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/MemoryMarshal/CreateSpan.cs]: Same here. Not sure why it was DangerousCreate1 to begin with.

File: src/System.Memory/tests/System.Memory.Tests.csproj [hunk 4]
```diff
@@ -4,9 +4,20 @@
   <PropertyGroup>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <ProjectGuid>{15DB0DCC-68B4-4CFB-8BD2-F26BCCAF5A3F}</ProjectGuid>
+    <IsPartialFacadeAssembly Condition="'$(TargetGroup)' == 'netcoreapp' OR '$(TargetGroup)' == 'uap'">true</IsPartialFacadeAssembly>
+    <DefineConstants Condition="'$(IsPartialFacadeAssembly)' != 'true'">$(DefineConstants);netstandard;FEATURE_PORTABLE_SPAN</DefineConstants>
```
Comment (ahsonkhan) [inline src/System.Memory/tests/System.Memory.Tests.csproj]: These constants seem unnecessary (not used).

File: src/System.Memory/ref/System.Memory.cs [hunk 5]
```diff
@@ -124,8 +122,6 @@ public static partial class MemoryExtensions
         public int Length { get { throw null; } }
         public void CopyTo(System.Span<T> destination) { }
         [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
-        public static System.ReadOnlySpan<T> DangerousCreate(object obj, ref T objectData, int length) { throw null; }
```
Comment (ahsonkhan) [inline src/System.Memory/ref/System.Memory.cs]: All these changes need to be made to the System.Runtime ref as well.

File: src/System.Memory/tests/System.Memory.Tests.csproj (hunk 4 above)
Comment (ianhays) [reply src/System.Memory/tests/System.Memory.Tests.csproj]: Good point. I added them to keep the new tests in a single file and used IFDEFS but decided that was cumbersome so broke them up into group-specific files included based on IsPartialFacadeAssembly.

I'll remove the DefineConstants.

**Comment by ianhays** [comment]
@ahsonkhan thanks for the feedback! I addressed all of it.

**Comment by ianhays** [comment]
cc: @GrabYourPitchforks @KrzysztofCwalina

**Comment by ianhays** [comment]
Pinging @KrzysztofCwalina @GrabYourPitchforks @ahsonkhan @stephentoub for final feedback.

File: src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.Portable.cs [hunk 6]
```diff
@@ -47,5 +47,57 @@ public static partial class MemoryMarshal
...
+            where TFrom : struct
+            where TTo : struct
+        {
+            if (SpanHelpers.IsReferenceOrContainsReferences<TFrom>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TFrom));
+
+            if (SpanHelpers.IsReferenceOrContainsReferences<TTo>())
+                ThrowHelper.ThrowArgumentException_InvalidTypeWithPointersNotSupported(typeof(TTo));
+
+            int newLength = checked((int)((long)source.Length * Unsafe.SizeOf<TFrom>() / Unsafe.SizeOf<TTo>()));
```
Comment (KrzysztofCwalina) [inline src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.Portable.cs]: Should we check that TTo's size is more than zero? i.e. what happens if I cast from span of bytes to a span of empty structs?

File: src/System.Memory/tests/System.Memory.Tests.csproj [hunk 7]
```diff
@@ -125,6 +131,9 @@
     <Compile Include="MemoryMarshal\AsMemory.cs" />
     <Compile Include="MemoryMarshal\GetReference.cs" />
     <Compile Include="MemoryMarshal\TryGetArray.cs" />
+    <Compile Include="MemoryMarshal\CastReadOnlySpan.cs" />
```
Comment (ahsonkhan) [inline src/System.Memory/tests/System.Memory.Tests.csproj]: nit: sort order as alphabetical

**Comment by AndyAyersMS** [comment]
The checked multiply pattern in `Cast` generates some ugly code which can be avoided in many cases by doing some case analysis.

See notes over in dotnet/coreclr#15492.

If we're going to leave it as is, then I'd get rid of the aggressive inlining attribute.

File: src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.Portable.cs (hunk 6 above)
Comment (ianhays) [reply src/System.Memory/src/System/Runtime/InteropServices/MemoryMarshal.Portable.cs]: `Unsafe.SizeOf<TTo>` returns a nonzero value for all types.

Added this test:
```C#
        struct EmptyStruct { }

        [Fact]
        public static void CastSpanToEmptyStruct()
        {
            Span<uint> span = new Span<uint>(new uint[] { 1 });
            Span<EmptyStruct> emptyspan = MemoryMarshal.Cast<uint, EmptyStruct>(span);
            Assert.Equal(1, Unsafe.SizeOf<EmptyStruct>());
            Assert.Equal(4, emptyspan.Length);
        }
```

**Comment by ahsonkhan** [comment]
@ianhays, can you please maintain the commit history during PR review? It makes it easier to view only the diffs between commits.

File: src/System.Memory/src/System/Span.cs [hunk 8]
```diff
@@ -346,6 +328,12 @@ public override int GetHashCode()
             throw new NotSupportedException(SR.CannotCallGetHashCodeOnSpan);
         }
 
+        /// <summary>
+        /// Returns a <see cref="String"/> with the name of the type and the number of elements
+        /// </summary>
+        /// <returns>A <see cref="String"/> with the name of the type and the number of elements</returns>
+        public override string ToString() => $"System.Span[{Length}]";
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Span.cs]: Should this  be identical to the DebuggerDisplay string (i.e. should we include typeof(T).Name to ToString) - [SpanDebuggerDisplay](https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Span.cs#L128) / [ReadOnlySpanDebuggerDisplay](https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/ReadOnlySpan.cs#L124)?

Similarly, should we add the prefix "System.Span" to the DebuggerDisplay, with angle brackets around the type T?

For instance, I propose the following:
```C#
Span<int> span = new int[10];
string str = span.ToString(); //System.Span<int>[10]
```

**Comment by ianhays** [comment]
>can you please maintain the commit history during PR review? It makes it easier to view only the diffs between commits.

Apologies, I wasn't expecting any more feedback. That rebase was to remove the feedback commits pre-merge.

File: src/System.Memory/src/System/Span.cs (hunk 8 above)
Comment (ianhays) [reply src/System.Memory/src/System/Span.cs]: Sure, that would be useful.

Comment (ianhays) [reply src/System.Memory/src/System/Span.cs]: Updated with the new ToString/DebuggerDisplay string.

I'll post a PR to coreclr with the new strings to FastSpan along with the removal of the old NonPortableCast functions after this PR in corefx is merged.

**Comment by ianhays** [comment]
Thanks for the feedback! merged.

**Pull request merged**